File: tuner-android/app/src/test/kotlin/com/opentuner/SmoothingTest.kt
package com.opentuner

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test

class SmoothingTest {
    private fun string(
        cents: Float?,
        dir: String?,
        conf: Float,
    ) = StrumReport(listOf(TuningString(0, "E2", 82f, cents?.let { 82f }, cents, dir, conf)))

    @Test
    fun strumHoldsLastConfidentReadingThenDrops() {
        val sm = StrumSmoother(holdMs = 1000, minConfidence = 0.2f)
        assertEquals(3f, sm.update(string(3f, "sharp", 0.9f), 0)[0].centsOff)
        // Silent frame within the hold window keeps the last reading.
        assertEquals(3f, sm.update(string(null, null, 0f), 500)[0].centsOff)
        // Beyond the hold window the reading drops.
        assertNull(sm.update(string(null, null, 0f), 2000)[0].centsOff)
    }

    @Test
    fun strumIgnoresLowConfidence() {
        val sm = StrumSmoother(holdMs = 1000, minConfidence = 0.5f)
        assertNull(sm.update(string(3f, "sharp", 0.1f), 0)[0].centsOff)
    }

    private fun chord(name: String?) =
        ChordResult(
            candidates = name?.let { listOf(ChordCandidate(it, 0.9f, 0, "", emptyList())) } ?: emptyList(),
            best = name?.let { ChordCandidate(it, 0.9f, 0, "", emptyList()) },
            strings = emptyList(),
        )

    @Test
    fun chordDebouncesBeforeShowing() {
        val sm = ChordSmoother(debounceMs = 250, holdMs = 1000)
        assertNull(sm.update(chord("C"), 0)) // pending starts
        assertNull(sm.update(chord("C"), 100)) // not stable yet
        assertNotNull(sm.update(chord("C"), 300)) // stable → shown
    }

    @Test
    fun chordHoldsAcrossBriefGapThenExpires() {
        val sm = ChordSmoother(debounceMs = 0, holdMs = 1000)
        assertNull(sm.update(chord("C"), 0)) // pending
        assertNotNull(sm.update(chord("C"), 10)) // confirmed → shown
        assertNotNull(sm.update(chord(null), 500)) // held across gap
        assertNull(sm.update(chord(null), 1100)) // expired
    }
}

File: tuner-web/tests/smoothing.test.ts
import { describe, it, expect } from "vitest";
import { StrumSmoother, ChordSmoother } from "../src/smoothing";
import type { ChordResult, StrumReport } from "../src/types";

function strum(cents: number | null, dir: "flat" | "in_tune" | "sharp" | null, conf: number): StrumReport {
  return {
    strings: [
      {
        index: 0,
        name: "E2",
        targetHz: 82,
        detectedHz: cents == null ? null : 82,
        centsOff: cents,
        direction: dir,
        confidence: conf,
      },
    ],
  };
}

function chord(name: string | null): ChordResult {
  return {
    candidates: name ? [{ name, score: 0.9 }] : [],
    best: name ? { name, score: 0.9 } : null,
    strings: [],
  };
}

describe("StrumSmoother", () => {
  it("holds the last confident reading then drops it", () => {
    const sm = new StrumSmoother(1000, 0.2);
    expect(sm.update(strum(3, "sharp", 0.9), 0)[0].centsOff).toBe(3);
    expect(sm.update(strum(null, null, 0), 500)[0].centsOff).toBe(3); // held
    expect(sm.update(strum(null, null, 0), 2000)[0].centsOff).toBeNull(); // expired
  });

  it("ignores low-confidence readings", () => {
    const sm = new StrumSmoother(1000, 0.5);
    expect(sm.update(strum(3, "sharp", 0.1), 0)[0].centsOff).toBeNull();
  });
});

describe("ChordSmoother", () => {
  it("debounces before showing", () => {
    const sm = new ChordSmoother(250, 1000);
    expect(sm.update(chord("C"), 0)).toBeNull();
    expect(sm.update(chord("C"), 100)).toBeNull();
    expect(sm.update(chord("C"), 300)).not.toBeNull();
  });

  it("holds across a brief gap then expires", () => {
    const sm = new ChordSmoother(0, 1000);
    expect(sm.update(chord("C"), 0)).toBeNull();
    expect(sm.update(chord("C"), 10)).not.toBeNull();
    expect(sm.update(chord(null), 500)).not.toBeNull();
    expect(sm.update(chord(null), 1100)).toBeNull();
  });
});
